src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,553 @@
+        }
+        // Reset forward navigation stack if we are capturing a new event
+        jumpedPosStack = [];
+        if (captureTimer) {
+            window.clearTimeout(captureTimer);
+            captureTimer = null;
+        }
+        
+        // Ensure cursor activity has not happened because of arrow keys or edit
+        if (selectionObj.origin !== "+move" && (!window.event || (window.event && window.event.type !== "input"))) {
+            captureTimer = window.setTimeout(function () {
+                // Check if we have reached MAX_NAV_FRAMES_COUNT
+                // If yes, control overflow
+                if (jumpToPosStack.length === MAX_NAV_FRAMES_COUNT) {
+                    var navFrame = jumpToPosStack.splice(0, 1)[0];
+                    navFrame._clearMarkers();
+                }
+
+                currentEditPos = new NavigationFrame(event.target, selectionObj);
+                jumpToPosStack.push(currentEditPos);
+                _validateNavigationCmds();
+                activePosNotSynced = false;
+            }, NAV_FRAME_CAPTURE_LATENCY);
+        } else {
+            activePosNotSynced = true;
+        }
+    }
+    
+   /**
+    * Command handler to navigate backward
+    */
+    function _navigateBack() {
+        if (!jumpedPosStack.length) {
+            if (activePosNotSynced) {
+                currentEditPos = new NavigationFrame(EditorManager.getCurrentFullEditor(), {ranges: EditorManager.getCurrentFullEditor()._codeMirror.listSelections()});
+                jumpedPosStack.push(currentEditPos);
+            }
+        }
+        
+        var navFrame = jumpToPosStack.pop();
+        
+        // Check if the poped frame is the current active frame or doesn't have any valid marker information
+        // if true, jump again
+        if (navFrame && navFrame === currentEditPos) {
+            jumpedPosStack.push(navFrame);
+            CommandManager.execute(NAVIGATION_JUMP_BACK);
+            return;
+        } else if (navFrame && !navFrame._validateMarkers()) {
+            CommandManager.execute(NAVIGATION_JUMP_BACK);
+            return;
+        }
+        
+        if (navFrame) {
+            // We will check for the file existence now, if it doesn't exist we will jump back again
+            // but discard the popped frame as invalid.
+            _checkIfExist(navFrame).done(function () {
+                jumpedPosStack.push(navFrame);
+                navFrame.goTo();
+                currentEditPos = navFrame;
+            }).fail(function () {
+                CommandManager.execute(NAVIGATION_JUMP_BACK);
+            }).always(function () {
+                _validateNavigationCmds();
+            });
+        }
+    }
+    
+   /**
+    * Command handler to navigate forward
+    */
+    function _navigateForward() {
+        var navFrame = jumpedPosStack.pop();
+        
+        // Check if the poped frame is the current active frame or doesn't have any valid marker information
```

Comment:
We should have an early `return` here for the case when navFrame is null (`!navFrame`)